File: src/biotech_rag_assistant/retrieval.py
"""BM25 retrieval over approved/effective controlled-document chunks."""

from __future__ import annotations

import re
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field
from rank_bm25 import BM25Okapi

from biotech_rag_assistant.chunking import chunk_documents
from biotech_rag_assistant.models import DocumentChunk, RetrievalHit, SourceDocument

TOKEN_RE = re.compile(r"\w+")


class RetrievalConfig(BaseModel):
    """Shared retrieval settings for CLI, evaluation, and answer assembly."""

    model_config = ConfigDict(extra="forbid")

    method: Literal["bm25"] = "bm25"
    top_k: int = Field(default=5, ge=1)
    score_floor: float = Field(default=0.0, ge=0.0)
    min_top_coverage: float = Field(default=0.6, ge=0.0, le=1.0)

    def to_cli_record(self) -> dict[str, object]:
        return self.model_dump()


def tokenize(text: str) -> list[str]:
    """Lowercase word tokenizer used by the deterministic BM25 baseline."""
    return TOKEN_RE.findall(text.lower())


# Generic, domain-safe stopwords removed before measuring query-term coverage.
# Deliberately excludes controlled-document vocabulary (status, version, approved,
# document, procedure, record, quality, ...), which must count toward relevance.
STOPWORDS = frozenset(
    """
    a an the of for to and or in on is are was were be been being this that these those
    what which who whom how when where why do does did we you they it its our your their
    with without within into onto from by as at if then than so such not no can could should
    would will may might must have has had i me my them he she his her him us about over under
    """.split()
)


def content_terms(text: str) -> set[str]:
    """Distinct non-stopword query tokens used to measure relevance coverage."""
    return {token for token in tokenize(text) if token not in STOPWORDS and len(token) > 1}


def coverage(query_text: str, chunk_text: str) -> float:
    """Fraction of the query's distinctive terms that appear in a chunk (0.0-1.0)."""
    query_terms = content_terms(query_text)
    if not query_terms:
        return 0.0
    chunk_tokens = set(tokenize(chunk_text))
    return len(query_terms & chunk_tokens) / len(query_terms)


def index_text(chunk: DocumentChunk) -> str:
    r"""The string indexed and relevance-gated for a chunk: section heading + verbatim span.

    ``chunk.text`` is the exact cited source span (``raw[char_start:char_end]``, ADR-016). The
    section heading is folded in only here, at index/scoring time, so heading terms stay
    matchable for BM25 and the coverage gate without the cited quote disagreeing with its own
    offsets. The result is byte-identical to the pre-ADR-016 ``heading + "\n" + body`` text, so
    BM25 ranking and the 0.6 coverage calibration are unchanged.
    """
    return f"{chunk.section_heading}\n{chunk.text}"


class BM25Retriever:
    """Thin wrapper around rank-bm25 with deterministic tie-breaking."""

    def __init__(self, chunks: list[DocumentChunk]) -> None:
        self.chunks = list(chunks)
        self._corpus_tokens = [tokenize(index_text(chunk)) for chunk in self.chunks]
        self._bm25 = BM25Okapi(self._corpus_tokens) if self.chunks else None

    def query(
        self,
        query_text: str,
        *,
        top_k: int = 5,
        score_floor: float = 0.0,
        min_top_coverage: float = 0.0,
    ) -> list[RetrievalHit]:
        query_tokens = tokenize(query_text)
        if not query_tokens or self._bm25 is None:
            return []

        scores = self._bm25.get_scores(query_tokens)
        ranked_indices = sorted(
            range(len(self.chunks)),
            key=lambda index: (
                -float(scores[index]),
                self.chunks[index].doc_id,
                self.chunks[index].chunk_index,
                self.chunks[index].chunk_id,
            ),
        )

        hits: list[RetrievalHit] = []
        for index in ranked_indices:
            score = float(scores[index])
            if score <= score_floor:
                continue
            hits.append(
                RetrievalHit(
                    rank=len(hits) + 1,
                    score=score,
                    chunk=self.chunks[index],
                )
            )
            if len(hits) >= top_k:
                break

        # Lexical relevance gate (ADR-012). BM25 returns positive scores for incidental
        # token overlap, so an off-topic question can otherwise surface a loosely related
        # approved passage that the extractive answer copies verbatim with valid citations.
        # Require the top-ranked passage to actually contain enough of the question's
        # distinctive terms; otherwise refuse (return no hits). Gating on rank 1 keeps the
        # decision independent of top_k and fails toward refusal.
        if hits and min_top_coverage > 0.0:
            if coverage(query_text, index_text(hits[0].chunk)) < min_top_coverage:
                return []
        return hits


def build_retriever(documents: list[SourceDocument], *, method: str = "bm25") -> BM25Retriever:
    """Chunk and index a corpus once into a reusable retriever.

    Building the BM25 index is the expensive step, and it depends only on the corpus, not on the
    query. Build it once per corpus (at API startup, or once before an evaluation loop) and reuse
    it via ``query_retriever``; only genuinely one-shot callers should use ``run_retrieval``.
    """
    if method != "bm25":
        raise ValueError(f"unsupported retrieval method: {method}")
    return BM25Retriever(chunk_documents(documents))


def query_retriever(
    retriever: BM25Retriever,
    query_text: str,
    config: RetrievalConfig,
) -> list[RetrievalHit]:
    """Query a prebuilt retriever with a config's query-time parameters."""
    return retriever.query(
        query_text,
        top_k=config.top_k,
        score_floor=config.score_floor,
        min_top_coverage=config.min_top_coverage,
    )


def run_retrieval(
    documents: list[SourceDocument],
    query_text: str,
    config: RetrievalConfig,
) -> list[RetrievalHit]:
    """One-shot convenience: build a retriever for ``documents`` and run a single query.

    For repeated queries over the same corpus, build once with ``build_retriever`` and reuse it
    with ``query_retriever`` rather than paying the chunk + index cost on every call.
    """
    return query_retriever(build_retriever(documents, method=config.method), query_text, config)

File: docs/retrieval.js
/*
 * Client-side port of the Biotech RAG Assistant deterministic retrieval path.
 * Mirrors the Python core (tokenize -> BM25Okapi -> score_floor -> deterministic
 * tie-break -> ADR-012 coverage relevance gate -> extractive answer). Runs in the
 * browser (static GitHub Pages demo) and under node (parity test). No backend.
 *
 * Faithful to rank-bm25 BM25Okapi defaults: k1=1.5, b=0.75, epsilon=0.25.
 * The corpus chunks are status-gated (Approved/Effective only) at export time,
 * exactly as Python chunk_documents() does, so Draft/Obsolete never appear here.
 */
(function (global) {
  "use strict";

  var STOPWORDS = new Set(
    ("a an the of for to and or in on is are was were be been being this that these those " +
      "what which who whom how when where why do does did we you they it its our your their " +
      "with without within into onto from by as at if then than so such not no can could should " +
      "would will may might must have has had i me my them he she his her him us about over under")
      .split(/\s+/)
  );

  function tokenize(text) {
    var m = String(text).toLowerCase().match(/\w+/g);
    return m || [];
  }

  function contentTerms(text) {
    var s = new Set();
    tokenize(text).forEach(function (t) {
      if (!STOPWORDS.has(t) && t.length > 1) s.add(t);
    });
    return s;
  }

  function coverage(queryText, chunkText) {
    var q = contentTerms(queryText);
    if (q.size === 0) return 0.0;
    var c = new Set(tokenize(chunkText));
    var hit = 0;
    q.forEach(function (t) { if (c.has(t)) hit += 1; });
    return hit / q.size;
  }

  // Faithful BM25Okapi (rank-bm25 defaults).
  function BM25(docsTokens, k1, b, epsilon) {
    this.k1 = k1 == null ? 1.5 : k1;
    this.b = b == null ? 0.75 : b;
    this.epsilon = epsilon == null ? 0.25 : epsilon;
    this.N = docsTokens.length;
    this.docFreqs = [];
    this.docLen = [];
    var nd = Object.create(null);
    var totalLen = 0;
    for (var i = 0; i < this.N; i++) {
      var tokens = docsTokens[i];
      this.docLen.push(tokens.length);
      totalLen += tokens.length;
      var freq = Object.create(null);
      for (var j = 0; j < tokens.length; j++) {
        var w = tokens[j];
        freq[w] = (freq[w] || 0) + 1;
      }
      this.docFreqs.push(freq);
      for (var word in freq) nd[word] = (nd[word] || 0) + 1;
    }
    this.avgdl = this.N ? totalLen / this.N : 0;
    this._calcIdf(nd);
  }

  BM25.prototype._calcIdf = function (nd) {
    this.idf = Object.create(null);
    var idfSum = 0;
    var negatives = [];
    var count = 0;
    for (var word in nd) {
      var freq = nd[word];
      var idf = Math.log(this.N - freq + 0.5) - Math.log(freq + 0.5);
      this.idf[word] = idf;
      idfSum += idf;
      count += 1;
      if (idf < 0) negatives.push(word);
    }
    this.averageIdf = count ? idfSum / count : 0;
    var eps = this.epsilon * this.averageIdf;
    for (var n = 0; n < negatives.length; n++) this.idf[negatives[n]] = eps;
  };

  BM25.prototype.getScores = function (queryTokens) {
    var scores = new Array(this.N).fill(0);
    for (var t = 0; t < queryTokens.length; t++) {
      var q = queryTokens[t];
      var idf = this.idf[q] || 0;
      if (idf === 0) continue;
      for (var i = 0; i < this.N; i++) {
        var f = this.docFreqs[i][q] || 0;
        if (f === 0) continue;
        var denom = f + this.k1 * (1 - this.b + (this.b * this.docLen[i]) / this.avgdl);
        scores[i] += idf * ((f * (this.k1 + 1)) / denom);
      }
    }
    return scores;
  };

  // engine over the exported corpus data; returns the same shape the UI/API renders.
  // The string indexed/gated for a chunk: section heading + verbatim span. Mirrors the Python
  // retrieval.index_text() (ADR-016) so JS BM25 and the coverage gate stay byte-identical to the
  // core even though chunk.text is now the exact cited source span (heading lives in its field).
  function indexText(c) { return c.section_heading + "\n" + c.text; }

  function makeEngine(data) {
    var chunks = data.chunks;
    var bm25 = new BM25(chunks.map(function (c) { return tokenize(indexText(c)); }));
    var minTopCoverage = data.min_top_coverage == null ? 0.6 : data.min_top_coverage;
    var scoreFloor = data.score_floor == null ? 0.0 : data.score_floor;

    function query(queryText, topK) {
      topK = topK || 3;
      var qTokens = tokenize(queryText);
      if (qTokens.length === 0) return [];
      var scores = bm25.getScores(qTokens);
      var order = chunks.map(function (_, i) { return i; });
      order.sort(function (a, b) {
        if (scores[a] !== scores[b]) return scores[b] - scores[a];        // -score
        if (chunks[a].doc_id !== chunks[b].doc_id) return chunks[a].doc_id < chunks[b].doc_id ? -1 : 1;
        if (chunks[a].chunk_index !== chunks[b].chunk_index) return chunks[a].chunk_index - chunks[b].chunk_index;
        return chunks[a].chunk_id < chunks[b].chunk_id ? -1 : 1;
      });
      var hits = [];
      for (var k = 0; k < order.length; k++) {
        var idx = order[k];
        if (scores[idx] <= scoreFloor) continue;
        hits.push({ rank: hits.length + 1, score: scores[idx], chunk: chunks[idx] });
        if (hits.length >= topK) break;
      }
      // ADR-012 relevance gate on the rank-1 hit.
      if (hits.length && minTopCoverage > 0.0) {
        if (coverage(queryText, indexText(hits[0].chunk)) < minTopCoverage) return [];
      }
      return hits;
    }

    function answer(queryText, topK) {
      var hits = query(queryText, topK);
      if (!hits.length) {
        return { outcome: "refusal", answer_text: data.refusal_text, hits: [],
                 review_recommended: true, reasons: ["refusal_no_supporting_documents"],
                 documents_excluded: data.documents_excluded, documents_retrievable: data.documents_retrievable };
      }
      return { outcome: "answer", answer_text: null,
               hits: hits.map(function (h) { return { score: h.score, chunk: h.chunk }; }),
               citations_valid: true, review_recommended: false, reasons: [],
               documents_excluded: data.documents_excluded, documents_retrievable: data.documents_retrievable };
    }

    return { query: query, answer: answer, bm25: bm25 };
  }

  var api = { tokenize: tokenize, contentTerms: contentTerms, coverage: coverage, BM25: BM25, makeEngine: makeEngine };
  if (typeof module !== "undefined" && module.exports) module.exports = api;
  else global.BiotechRAG = api;
})(typeof window !== "undefined" ? window : globalThis);
